internal/service/federateddatabaseinstance/resource_schema.go

```diff
@@ -0,0 +1,86 @@
+package federateddatabaseinstance
+
+import "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+func cloudProviderConfig(isDataSource bool) *schema.Schema {
+	var computed, optional, required bool
+	var maxItems int
+	if isDataSource {
+		computed = true
+		maxItems = 0
+	} else {
+		required = true
+		optional = true
+		maxItems = 1
+	}
+
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		MaxItems: maxItems,
+		Computed: true,
+		Optional: true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"aws": {
+					Type:     schema.TypeList,
+					MaxItems: maxItems,
+					Optional: true,
+					Computed: computed,
+					Elem: &schema.Resource{
+						Schema: map[string]*schema.Schema{
+							"role_id": {
+								Type:     schema.TypeString,
+								Required: required,
+								Computed: computed,
+							},
+							"test_s3_bucket": {
+								Type:     schema.TypeString,
+								Required: required,
+								Optional: isDataSource,
+							},
+							"iam_assumed_role_arn": {
+								Type:     schema.TypeString,
+								Computed: true,
+							},
+							"iam_user_arn": {
+								Type:     schema.TypeString,
+								Computed: true,
+							},
+							"external_id": {
+								Type:     schema.TypeString,
+								Computed: true,
+							},
+						},
+					},
+				},
+				"azure": {
+					Type:     schema.TypeList,
+					MaxItems: maxItems,
+					Optional: optional,
+					Computed: computed,
+					Elem: &schema.Resource{
+						Schema: map[string]*schema.Schema{
+							"role_id": {
+								Type:     schema.TypeString,
+								Required: required,
+								Computed: computed,
+							},
```

Comment:
```suggestion
								Required: !isDataSource,
								Computed: isDataSource,
							},
```
nit: Having the `required`/`computed` variables adds one additional level of indirection, might be easier to follow if the criteria is explicit in the schema directly.

Reply:
I understand the concern about the level of indirection, however I believe for readability purposes it's better to keep the variables as they are, since it won't require a mental mapping of the boolean logic.